### PR TITLE
[Midend] Enhancements and Optimizations and [Examples] Added MLIRLinalg Examples for Various Optimization Options

### PR DESCRIPTION
**Pull request by 6somehow**
### [Midend] Enhancements and Optimizations

- **BatchMatMul:**
  - Updated unroll, tile, and vectorization strategies for BatchMatMul operations.
  - Enhanced SCF (Structured Control Flow) version for BatchMatMul to support further optimization and work.

- **Conv2D NHWC-FHWC:**
  - Improved vectorization and tiling techniques for Conv2D operations in NHWC-FHWC layout.
  - These updates provide a vectorization option and a tile+vectorization option for buddy-opt.

- **Depthwise Conv2D NHWC-HWC:**
  - Introduced vectorization and tiling strategies for Depthwise Conv2D in NHWC-HWC format, ensuring more efficient execution.

### [Examples] Added MLIRLinalg Examples for Various Optimization Options

New MLIRLinalg examples were added to demonstrate the usage of new buddy-opt various optimization options , including:
1. **conv-nhwc-fhwc-optimize:**  Conv2D NHWC-FHWC layout vectorization optimization.
2. **conv-nhwc-fhwc-tile-optimize:** Conv2D NHWC-FHWC vectorization  with tiling optimizations.
3. **depthwise-conv-nhwc-hwc-optimize:** Depthwise Conv2D NHWC-HWC 
vectorization  layout optimization.
4. **batchmatmul-tile-optimize:** tiling and unroll optimization for BatchMatMul.
5. **batchmatmul-scf-optimize:** using SCF optimization for BatchMatMul.

### Example MLIR Files:
- `batchmatmul`
- `conv2d_nhwc_fhwc`
- `depthwise_conv_2d_nhwc_hwc`

These updates collectively improve the performance of matrix and convolution operations in MLIR, providing optimized patterns for common workloads like BatchMatMul and convolution layers.

### CODE REVIEW

File: midend/lib/Conversion/CMakeLists.txt [hunk 1]
```diff
@@ -14,3 +14,4 @@ add_subdirectory(LowerLinalgToGemmini)
 add_subdirectory(SchedulingOnDevices)
 add_subdirectory(LowerSche)
 add_subdirectory(FuncBufferize)
+add_subdirectory(DepthwiseConvOptimization)
```
Comment (xlinsist) [inline midend/lib/Conversion/CMakeLists.txt:17]: Add an empty blank in the end of the file.

File: llvm
Comment (xlinsist) [inline llvm:1]: Remove the modification of LLVM.

File: examples/MLIRLinalg/linalg-depthwise_conv_2d_nhwc_hwc.mlir [hunk 2]
```diff
@@ -0,0 +1,7 @@
+func.func @depthwise_conv_2d_nhwc_hwc(%arg0: memref<?x?x?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?x?x?xf32>) {
```
Comment (xlinsist) [inline examples/MLIRLinalg/linalg-depthwise_conv_2d_nhwc_hwc.mlir:1]: Add lit' s file check.

File: examples/MLIRLinalg/makefile [hunk 3]
```diff
@@ -60,6 +60,24 @@ linalg-conv2d-tiling-run:
 		-convert-func-to-llvm -reconcile-unrealized-casts | \
 	${MLIR_CPU_RUNNER} ${OPT_FLAG} -e main -entry-point-result=void -shared-libs=${MLIR_RUNNER_UTILS} -shared-libs=${MLIR_C_RUNNER_UTILS}
 
+linalg-conv2d_nhwc_fhwc-optimize-lower:
+	@${BUDDY_OPT} linalg-conv2d_nhwc_fhwc.mlir \
+		--conv-nhwc-fhwc-optimize="vec-size=16" \
```
Comment (xlinsist) [inline examples/MLIRLinalg/makefile:65]: Should it be `--conv-nhwc-fhwc-tile-optimize`?

File: examples/ConvOpt/pointwise_conv2d_nhwc_filter_hwcf.mlir [hunk 4]
```diff
@@ -1,9 +1,6 @@
 // Generated from Mobilenet.mlir file
 func.func @conv_2d_1x1(%input: tensor<1x4x5x2xf32>, %filter: tensor<1x1x2x7xf32>) -> tensor<1x4x5x7xf32> {
     %0 = linalg.init_tensor [1, 4, 5, 7] : tensor<1x4x5x7xf32>
-    %1 = linalg.conv_2d_nhwc_hwcf {
```
Comment (xlinsist) [inline examples/ConvOpt/pointwise_conv2d_nhwc_filter_hwcf.mlir:4]: What's the purpose of this modification?

File: midend/lib/Conversion/ConvOptimization/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,5 @@
 add_mlir_library(ConvOptimization
 	ConvOptimize.cpp
+  ConvNhwcFhwcOptimize.cpp
+  ConvNhwcFhwcOptimizeTile.cpp
```
Comment (xlinsist) [inline midend/lib/Conversion/ConvOptimization/CMakeLists.txt:4]: What is the difference between `ConvNhwcFhwcOptimizeTile` and `ConvNhwcFhwcOptimize`? Do they have different usages or applications? Or is `ConvNhwcFhwcOptimizeTile` always better than `ConvNhwcFhwcOptimize`?

File: midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimize.cpp [hunk 6]
```diff
@@ -0,0 +1,278 @@
+//====- ConvNhwcFhwcOptimize.cpp
```
Comment (xlinsist) [inline midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimize.cpp:1]: Recommand to centralize the title.

File: midend/lib/Conversion/MatMulOptimization/CMakeLists.txt [hunk 7]
```diff
@@ -11,6 +13,12 @@ add_mlir_library(BatchMatMulOptimization
   BatchMatMulOptimize.cpp
 )
 
+add_mlir_library(BatchMatMulTileOptimization
+  BatchMatMulTileOptimize.cpp
+  BatchMatMulSCFOptimize.cpp
```
Comment (xlinsist) [inline midend/lib/Conversion/MatMulOptimization/CMakeLists.txt:18]: Why is `BatchMatMulSCFOptimize` under `BatchMatMulTileOptimization`？

File: tools/buddy-opt/buddy-opt.cpp [hunk 8]
```diff
@@ -40,31 +40,37 @@
...
 #include "Sche/ScheDialect.h"
 #include "Sche/ScheOps.h"
+#include "VectorExp/VectorExpDialect.h"
+#include "VectorExp/VectorExpOps.h"
 
 namespace mlir {
 namespace buddy {
 void registerConvVectorizationPass();
 void registerPointwiseConvToGemmPass();
+// void registerPointwiseConvToGemmForNhwcFhwcPass();
```
Comment (xlinsist) [inline tools/buddy-opt/buddy-opt.cpp:55]: Could we uncomment it?

File: tools/buddy-opt/CMakeLists.txt [hunk 9]
```diff
@@ -26,9 +26,11 @@ target_link_libraries(buddy-opt
   LowerRVVPass
   MatMulOptimization
   BatchMatMulOptimization
+  BatchMatMulTileOptimization
```
Comment (xlinsist) [inline tools/buddy-opt/CMakeLists.txt:29]: Can we unify `BatchMatMulOptimization` and `BatchMatMulTileOptimization`?

File: examples/MLIRLinalg/makefile (hunk 3 above)
Comment (6somehow) [reply examples/MLIRLinalg/makefile:65]: Due to the uncertainty regarding the benefits of tiling, both --conv-nhwc-fhwc-tile-optimize and --conv-nhwc-fhwc-optimize options are available.

File: examples/ConvOpt/pointwise_conv2d_nhwc_filter_hwcf.mlir (hunk 4 above)
Comment (6somehow) [reply examples/ConvOpt/pointwise_conv2d_nhwc_filter_hwcf.mlir:4]: This modification will be undone.

File: midend/lib/Conversion/ConvOptimization/CMakeLists.txt (hunk 5 above)
Comment (6somehow) [reply midend/lib/Conversion/ConvOptimization/CMakeLists.txt:4]: The benefit of tiling depends on the input data size and the hardware cache size.

File: tools/buddy-opt/buddy-opt.cpp (hunk 8 above)
Comment (6somehow) [reply tools/buddy-opt/buddy-opt.cpp:55]: Sure

File: tools/buddy-opt/CMakeLists.txt (hunk 9 above)
Comment (6somehow) [reply tools/buddy-opt/CMakeLists.txt:29]: There are two technical approaches to achieve tiling in batch matmul. One is similar to BatchMatMulTileOptimization, which uses subviews for tiling. The other is to adapt scf.forall (as in BatchMatMulSCFOptimization) through transformations in MLIR. This is subject to further discussion.

File: midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimizeTile.cpp [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+
+    Value stepOH = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOH)), OH);
+    Value stepOW = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOW)), OW);
+    Value stepOC = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOC)), OC);
+
+    // clang format off
+    //  Step 1: Create outer most loops.
```
Comment (xlinsist) [inline midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimizeTile.cpp:122]: Do we have `Step 2` in this file? If not, the word `Step` had better be removed.

File: midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimize.cpp [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    Value N = rewriter.create<memref::DimOp>(loc, output, 0);  // N
+    Value OH = rewriter.create<memref::DimOp>(loc, output, 1); // OH
+    Value OW = rewriter.create<memref::DimOp>(loc, output, 2); // OW
+    Value OC = rewriter.create<memref::DimOp>(loc, output, 3); // OC
+    Value IC = rewriter.create<memref::DimOp>(loc, input, 3);  // IC
+    Value FH = rewriter.create<memref::DimOp>(loc, filter, 1); // FH
+    Value FW = rewriter.create<memref::DimOp>(loc, filter, 2); // FW
+
+    // clang format off
+    //  Step 1: Create outer most loops.
```
Comment (xlinsist) [inline midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimize.cpp:106]: Do we have `Step 2` in this file? If not, the word `Step` had better be removed.

File: examples/MLIRLinalg/makefile [hunk 12]
```diff
@@ -60,6 +60,37 @@ linalg-conv2d-tiling-run:
...
+		-convert-func-to-llvm -reconcile-unrealized-casts | \
+	${MLIR_CPU_RUNNER} ${OPT_FLAG} -e main -entry-point-result=void -shared-libs=${MLIR_RUNNER_UTILS} -shared-libs=${MLIR_C_RUNNER_UTILS}
+
+linalg-conv2d_nhwc_fhwc-tile-optimize-run:
+	@${BUDDY_OPT} linalg-conv2d_nhwc_fhwc.mlir ${MLIR_OPT_OPTIONS} \
+		-conv-nhwc-fhwc-tile-optimize="vec-size=16 tiling-height=2 tiling-width=3" \
+		-lower-affine -convert-scf-to-cf \
+		-convert-vector-to-llvm -finalize-memref-to-llvm -convert-arith-to-llvm \
+		-convert-func-to-llvm -reconcile-unrealized-casts | \
+	${MLIR_CPU_RUNNER} ${OPT_FLAG} -e main -entry-point-result=void -shared-libs=${MLIR_RUNNER_UTILS} -shared-libs=${MLIR_C_RUNNER_UTILS}
```
Comment (xlinsist) [inline examples/MLIRLinalg/makefile:87]: Better to obey the order of the existing test methods. For example, the order of tests in this code should be: 
1. linalg-conv2d_nhwc_fhwc-optimize-lower
2. linalg-conv2d_nhwc_fhwc-optimize-run
3. linalg-conv2d_nhwc_fhwc-tile-optimize-lower
4. linalg-conv2d_nhwc_fhwc-tile-optimize-run

File: midend/lib/Conversion/ConvOptimization/ConvOptimize.cpp [hunk 13]
```diff
@@ -63,88 +69,198 @@ class ConvOptimizePattern : public ConversionPattern {
...
+                            Value bias = builder.create<memref::LoadOp>(
+                                loc, output, ValueRange{ivA, ivB, ivC, ivD});
+                            Value addRes = builder.create<arith::AddFOp>(
+                                loc, bias, reducedRes);
+                            builder.create<memref::StoreOp>(
+                                loc, addRes, output,
+                                ValueRange{ivA, ivB, ivC, ivD});
+                          });
+                    });
               });
```
Comment (xlinsist) [inline midend/lib/Conversion/ConvOptimization/ConvOptimize.cpp:262]: Is the purpose of modifying this code for beautifying/formatting? Would the original code look better?

File: midend/lib/Conversion/DepthwiseConvOptimization/DepthwiseConvNhwcHwc.cpp [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+        loc, AffineMap::get(1, 0, d0 % vecSize), ValueRange{OC});
+    Value maskVector = rewriter.create<vector::CreateMaskOp>(
+        loc, VectorType::get({vecSize}, rewriter.getI1Type()),
+        ValueRange{tailLength});
+
+    Value FH = rewriter.create<memref::DimOp>(loc, filter, 0); // FH
+    Value FW = rewriter.create<memref::DimOp>(loc, filter, 1); // FW
+
+    // clang format off
+    //  Step 1: Create outer most loops.
```
Comment (xlinsist) [inline midend/lib/Conversion/DepthwiseConvOptimization/DepthwiseConvNhwcHwc.cpp:123]: Do we have `Step 2` in this file? If not, the word `Step` had better be removed.

File: midend/lib/Conversion/MatMulOptimization/BatchMatMulOptimize.cpp [hunk 15]
```diff
@@ -81,16 +81,28 @@ class BatchMatMulOptimizePattern : public ConversionPattern {
...
+            ? rewriter
+                  .create<vector::BroadcastOp>(
+                      loc, VectorType::get({affineVectorSize}, elementType),
+                      zeroElementType)
+                  .getResult()
+            : rewriter
+                  .create<vector::SplatOp>(
+                      loc, VectorType::get({affineVectorSize}, elementType),
+                      zeroElementType)
+                  .getResult();
```
Comment (xlinsist) [inline midend/lib/Conversion/MatMulOptimization/BatchMatMulOptimize.cpp:96]: What is the reason for this change(If there are actual changes, there is no need to add comments in this code, I just wonder)?

File: midend/lib/Conversion/MatMulOptimization/CMakeLists.txt [hunk 16]
```diff
@@ -11,6 +13,14 @@ add_mlir_library(BatchMatMulOptimization
   BatchMatMulOptimize.cpp
 )
 
+add_mlir_library(BatchMatMulTileOptimization
+  BatchMatMulTileOptimize.cpp
+)
+
+add_mlir_library(BatchMatMulSCFOptimization
+  BatchMatMulSCFOptimize.cpp
+)
```
Comment (xlinsist) [inline midend/lib/Conversion/MatMulOptimization/CMakeLists.txt:22]: Do `BatchMatMulOptimize`, `BatchMatMulTileOptimize` and `BatchMatMulSCFOptimize` need to "add_mlir_library" twice? Should all three of them belong to `BatchMatMulOptimization`?

**Comment by xlinsist** [comment]
It seems that the FileCheck of mlir files in examples/MLIRLinalg directory does not pass the lit check. When you have modified this part, you need to use `ninja check-buddy` command in the build folder to check whether the modified part is correct.

image

**Comment by xlinsist** [comment]
Running `make linalg-depthwise_conv_2d_nhwc_hwc-optimize-lower` leads to lowering error:

image

**Comment by xlinsist** [comment]
Running either `linalg-batch-matmul-tile-optimize-lower` or  `linalg-batch-matmul-scf-optimize-lower` leads to the following error:

image

**Review by xlinsist — CHANGES_REQUESTED** [review]
When I try to understand the optimization passes of this PR, I have some questions that I would like to verify:

1. We don't need to modify the strategies of the original ConvOptimize.cpp and BatchMatMulOptimize pass implementations, right? But I saw that there are code changes in the PR.

2. `linalg-conv2d_nhwc_fhwc-optimize` vectorizes the Channel dimension, and the number of vector elements in one iteration is fixed to 16, with no tail processing implemented. Does this mean this pass is only applicable to scenarios where the number of channels is divisible by 16?

3. The block size of `linalg-conv2d_nhwc_fhwc-tile-optimize` is 2x3, which I think should be the block size for Height and Width? It would be better if there are comments to elaborate optimization strategies in the code. 

4. The lowerings of `linalg-batch-matmul-scf-optimize`, `linalg-batch-matmul-tile-optimize`, `linalg-depthwise_conv_2d_nhwc_hwc-optimize` in makefile report error. It is recommended to ensure that the examples in examples/MLIRLinalg can be generated correctly, before integrating these passes into buddy-benchmark for Ops testing.
